src/script.rs

```diff
@@ -0,0 +1,589 @@
+            result.push(if neg { 0x80 } else { 0 });
+        } else if neg {
+            if let Some(last) = result.last_mut() {
+                *last |= 0x80;
+            }
+        }
+
+        result
+    }
+
+    fn set_vch(vch: &Vec<u8>) -> Result<i64, ScriptNumError> {
+        match vch.last() {
+            None => Ok(0),
+            Some(vch_back) => {
+                if *vch == vec![0, 0, 0, 0, 0, 0, 0, 128, 128] {
+                    // On an x86_64 system, the code below would actually decode the buggy
+                    // INT64_MIN encoding correctly. However in this case, it would be
+                    // performing left shifts of a signed type by 64, which has undefined
+                    // behavior.
+                    return Ok(i64::MIN);
+                };
+
+                // Guard against undefined behavior. INT64_MIN is the only allowed 9-byte encoding.
+                if vch.len() > 8 {
+                    return Err(ScriptNumError::Overflow {
+                        max_num_size: 8,
+                        actual: vch.len(),
+                    });
+                };
+
+                let mut result: i64 = 0;
+                for (i, vch_i) in vch.iter().enumerate() {
+                    result |= i64::from(*vch_i) << (8 * i);
+                }
+
+                // If the input vector's most significant byte is 0x80, remove it from
+                // the result's msb and return a negative.
+                if vch_back & 0x80 != 0 {
+                    return Ok(-(result & !(0x80 << (8 * (vch.len() - 1)))));
+                };
+
+                Ok(result)
+            }
+        }
+    }
+}
+
+impl Add for ScriptNum {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        let rhs = other.0;
+        assert!(
+            rhs == 0
+                || (rhs > 0 && self.0 <= i64::MAX - rhs)
+                || (rhs < 0 && self.0 >= i64::MIN - rhs)
+        );
+        Self(self.0 + rhs)
+    }
+}
+
+impl Sub for ScriptNum {
+    type Output = Self;
+
+    fn sub(self, other: Self) -> Self {
+        let rhs = other.0;
+        assert!(
+            rhs == 0
+                || (rhs > 0 && self.0 >= i64::MIN + rhs)
+                || (rhs < 0 && self.0 <= i64::MAX + rhs)
+        );
+        Self(self.0 - rhs)
+    }
+}
+
+impl Neg for ScriptNum {
+    type Output = Self;
+
+    fn neg(self) -> Self {
+        assert!(self.0 != i64::MIN);
+        Self(-self.0)
+    }
+}
+
+/** Serialized script, used inside transaction inputs and outputs */
+#[derive(Clone, Debug)]
+pub struct Script<'a>(pub &'a [u8]);
+
+impl Script<'_> {
+    pub fn get_op(script: &mut &[u8]) -> Result<Opcode, ScriptError> {
+        Self::get_op2(script, &mut vec![])
+    }
+
+    pub fn get_op2(script: &mut &[u8], buffer: &mut Vec<u8>) -> Result<Opcode, ScriptError> {
+        if script.is_empty() {
+            panic!("attempting to parse an opcode from an empty script");
+        }
```

Comment:
The original seems to simply return if that happens?

```
        if (pc >= end())
            return false;
```

Reply:
Good catch! I actually have a future commit that refactors this function to avoid mutation, and I was careful to preserve this `panic!`, because I thought it was part of the original 🤦.

I will gladly fix this!

**NB**: I think this `panic!` is unreachable in the current implementation, because all the call sites check first (in order to terminate iteration), but that’s no excuse.